// remotion/dist/cjs/Composition.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ClipComposition = exports.Composition = void 0;
const jsx_runtime_1 = require("react/jsx-runtime");
const react_1 = require("react");
const react_dom_1 = require("react-dom");
const AbsoluteFill_js_1 = require("./AbsoluteFill.js");
const CanUseRemotionHooks_js_1 = require("./CanUseRemotionHooks.js");
const CompositionManager_js_1 = require("./CompositionManager.js");
const input_props_js_1 = require("./config/input-props.js");
const delay_render_js_1 = require("./delay-render.js");
const Folder_js_1 = require("./Folder.js");
const get_environment_js_1 = require("./get-environment.js");
const internals_js_1 = require("./internals.js");
const loading_indicator_js_1 = require("./loading-indicator.js");
const NativeLayers_js_1 = require("./NativeLayers.js");
const nonce_js_1 = require("./nonce.js");
const portal_node_js_1 = require("./portal-node.js");
const use_lazy_component_js_1 = require("./use-lazy-component.js");
const use_video_js_1 = require("./use-video.js");
const validate_composition_id_js_1 = require("./validation/validate-composition-id.js");
const validate_dimensions_js_1 = require("./validation/validate-dimensions.js");
const validate_duration_in_frames_js_1 = require("./validation/validate-duration-in-frames.js");
const validate_fps_js_1 = require("./validation/validate-fps.js");
const Fallback = () => {
    (0, react_1.useEffect)(() => {
        const fallback = (0, delay_render_js_1.delayRender)('Waiting for Root component to unsuspend');
        return () => (0, delay_render_js_1.continueRender)(fallback);
    }, []);
    return null;
};
/**
 * @description This component is used to register a video to make it renderable and make it show in the sidebar, in dev mode.
 * @see [Documentation](https://www.remotion.dev/docs/composition)
 */
const Composition = ({ width, height, fps, durationInFrames, id, defaultProps, ...compProps }) => {
    const { registerComposition, unregisterComposition } = (0, react_1.useContext)(CompositionManager_js_1.CompositionManager);
    const video = (0, use_video_js_1.useVideo)();
    const lazy = (0, use_lazy_component_js_1.useLazyComponent)(compProps);
    const nonce = (0, nonce_js_1.useNonce)();
    const environment = (0, get_environment_js_1.useRemotionEnvironment)();
    const canUseComposition = (0, react_1.useContext)(internals_js_1.Internals.CanUseRemotionHooks);
    if (canUseComposition) {
        if (environment === 'player-development' ||
            environment === 'player-production') {
            throw new Error('<Composition> was mounted inside the `component` that was passed to the <Player>. See https://remotion.dev/docs/wrong-composition-mount for help.');
        }
        throw new Error('<Composition> mounted inside another composition. See https://remotion.dev/docs/wrong-composition-mount for help.');
    }
    const { folderName, parentName } = (0, react_1.useContext)(Folder_js_1.FolderContext);
    (0, react_1.useEffect)(() => {
        // Ensure it's a URL safe id
        if (!id) {
            throw new Error('No id for composition passed.');
        }
        (0, validate_composition_id_js_1.validateCompositionId)(id);
        (0, validate_dimensions_js_1.validateDimension)(width, 'width', 'of the <Composition/> component');
        (0, validate_dimensions_js_1.validateDimension)(height, 'height', 'of the <Composition/> component');
        (0, validate_duration_in_frames_js_1.validateDurationInFrames)({
            durationInFrames,
            component: 'of the <Composition/> component',
            allowFloats: false,
        });
        (0, validate_fps_js_1.validateFps)(fps, 'as a prop of the <Composition/> component', false);
        registerComposition({
            durationInFrames,
            fps,
            height,
            width,
            id,
            folderName,
            component: lazy,
            defaultProps,
            nonce,
            parentFolderName: parentName,
        });
        return () => {
            unregisterComposition(id);
        };
    }, [
        durationInFrames,
        fps,
        height,
        lazy,
        id,
        folderName,
        defaultProps,
        registerComposition,
        unregisterComposition,
        width,
        nonce,
        parentName,
    ]);
    if (environment === 'preview' && video && video.component === lazy) {
        const Comp = lazy;
        const inputProps = (0, input_props_js_1.getInputProps)();
        return (0, react_dom_1.createPortal)((0, jsx_runtime_1.jsx)(exports.ClipComposition, { children: (0, jsx_runtime_1.jsx)(CanUseRemotionHooks_js_1.CanUseRemotionHooksProvider, { children: (0, jsx_runtime_1.jsx)(react_1.Suspense, { fallback: (0, jsx_runtime_1.jsx)(loading_indicator_js_1.Loading, {}), children: (0, jsx_runtime_1.jsx)(Comp, { ...defaultProps, ...inputProps }) }) }) }), (0, portal_node_js_1.portalNode)());
    }
    if (environment === 'rendering' && video && video.component === lazy) {
        const Comp = lazy;
        const inputProps = (0, input_props_js_1.getInputProps)();
        return (0, react_dom_1.createPortal)((0, jsx_runtime_1.jsx)(CanUseRemotionHooks_js_1.CanUseRemotionHooksProvider, { children: (0, jsx_runtime_1.jsx)(react_1.Suspense, { fallback: (0, jsx_runtime_1.jsx)(Fallback, {}), children: (0, jsx_runtime_1.jsx)(Comp, { ...defaultProps, ...inputProps }) }) }), (0, portal_node_js_1.portalNode)());
    }
    return null;
};
exports.Composition = Composition;
const ClipComposition = ({ children }) => {
    const { clipRegion } = (0, react_1.useContext)(NativeLayers_js_1.NativeLayersContext);
    const style = (0, react_1.useMemo)(() => {
        return {
            display: 'flex',
            flexDirection: 'row',
            opacity: clipRegion === 'hide' ? 0 : 1,
            clipPath: clipRegion && clipRegion !== 'hide'
                ? `polygon(${clipRegion.x}px ${clipRegion.y}px, ${clipRegion.x}px ${clipRegion.height + clipRegion.y}px, ${clipRegion.width + clipRegion.x}px ${clipRegion.height + clipRegion.y}px, ${clipRegion.width + clipRegion.x}px ${clipRegion.y}px)`
                : undefined,
        };
    }, [clipRegion]);
    return (0, jsx_runtime_1.jsx)(AbsoluteFill_js_1.AbsoluteFill, { style: style, children: children });
};
exports.ClipComposition = ClipComposition;
